### ISSUE TITLE
Bug: Calling `Array.isArray` on type `object | object[]` reduces the array items to `any` instead of `object`

### ISSUE THREAD

**Issue by danielrozenberg** [typescript-eslint/typescript-eslint#5578]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Issue Description

Using `type X = object | object[]`, I expect code that follows from a true `Array.isArray(x)` to consider the items inside `x` to be of type `object`, however since TypeScript 4.8 it seems that `typescript-eslint` reduced the items' type to `any`:

e.g., this code works in TS 4.7 but fails in 4.8:
```ts
export function doSomething(obj: object | object[]): boolean {
  if (Array.isArray(obj)) {
    return obj.every((o) => doSomething(o));
  }
  return true;
}
```

With TS 4.8, eslint throws this error:
```
/home/daniel/minimal-ts-array-issue-reproduce/index.ts
  3:41  error  Unsafe argument of type `any` assigned to a parameter of type `object | object[]`  @typescript-eslint/no-unsafe-argument

✖ 1 problem (1 error, 0 warnings)
```

### Reproduction Repository Link

https://github.com/danielrozenberg/minimal-ts-array-issue-reproduce

### Repro Steps

1) clone the repo
2) `npm ci`
3) `npm test`

Switch between the `failing-example` (default) and the `working-example` branches to see the difference

### Versions

| package                                | version |
| -------------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin`     | `5.36.1` |
| `@typescript-eslint/parser`            | `5.36.1` |
| `@typescript-eslint/scope-manager`     | `unused` |
| `@typescript-eslint/typescript-estree` | `unused` |
| `@typescript-eslint/type-utils`        | `unused` |
| `@typescript-eslint/utils`             | `unused` |
| `TypeScript`                           | `4.8.2` |
| `ESLint`                               | `8.23.0` |
| `node`                                 | `16.17.0` |

**Comment by bradzacher** [comment]
This isn't actually a problem with our lint rule - our rule is working entirely as intended!

Instead this is a quirk that seems to have been introduced into TS v4.8.2.

If you open your code [in the TS playground](https://www.typescriptlang.org/play?ts=4.8.2#code/KYDwDg9gTgLgBAMwK4DsDGMCWEVwCYQDKEAtsDABaYoDmAFBAEYBWAXHE88BnAD4ctuMANoBdAJTtGECABtgAQ1wBvAFBw4mBHDoBBKFAUBPAHSYAzvsNGGLceLhqNGzgG51cAPSe4APQD8HhpBcFDkSFC4nCbAAG7AUDYMDgC8AHyOIS7uzt5+gVn4RKTkVLTJIQC+4jlwlR5hMBG4MFBIwO6VQA), I've annotated some inlays to highlight TS's resolved type for your variables.

You'll notice that after the `Array.isArray` call, TS believes that `obj` is of type `any[] | object[]`. So when you call `.every`, the type of the argument is `obj | any`, which TS just simplifies to `any`.

I have reported a TS bug on your behalf - https://github.com/microsoft/TypeScript/issues/50561

**Comment by bradzacher** [comment]
Looping back from the TS issue.

`any[]` is a subtype of `object`; which is why `Array.isArray` refines it to `any[]`. So this is working as intended from the TS side.

This if one of the reasons we recommend against using the `object` type - because it doesn't mean "an object", it means "any non-nullish type"!

**Comment by Josh-Cena** [comment]
@bradzacher `object` means "any non-primitive". `Object` and `{}` are "any non-nullish".

**Comment by bradzacher** [comment]
Oops yeah sorry. Mixed them up. But the rest of the comment is correct.

Unbanned because enough people complained :)
But still recommended against personally. Hard to refine out of and confusing meaning to most people!

**Comment by Josh-Cena** [comment]
TBH TS 4.8 did a lot of work around `{}` and I think it's time to teach people how to use it properly. Before we have negated types this is probably the best to have.

**Comment by danielrozenberg** [comment]
Thanks for looking into this! I didn't know about the `// ^?` comment on TS Playground, good to know that exists.

Unfortunately I'm using a library that defines a type as `type X = object | string | number; type Y = X | X[];`, so this ends up getting caught in this `Array.isArray` blunder. A workaround for us is using `x instanceof Array` instead, which doesn't expand the types

**Comment by bradzacher** [comment]
> A workaround for us is using x instanceof Array instead

There's a good reason that Array.isArray exists.
instanceof is not a good way to check as it can be inconsistent in many cases.

Dont change your code to satisfy TS here. Write correct code and add an assertion to overrule TS.

**Comment by alanorozco** [comment]
> @bradzacher instanceof is not a good way to check as it can be inconsistent in many cases.

The case that @danielrozenberg mentioned does not run on the browser, so we don't encounter the issue where `instanceof Array` fails because the reference to `Array` is mismatched across iframes. But I'm likely unaware of other cases, so I'm curious what those are.

**Comment by bradzacher** [comment]
The fact that it incorrectly refines the type is another point against it.

As mentioned - `object` is a super type of array, so the fact that `Array.isArray` refines `object` to `any[]` is correct!

Put another way - for the type `object`, it is safe to assign a `number[]`!

If `instanceof Array` doesn't refine your union to the same as `Array.isArray` then it's a bug in TS and introduces unsoundness in your code.

You should probably be guarding your code by using `typeof o === 'object' && o != null` to guard the element function against bad values.
